### PR TITLE
feat(vow): fix issues revealed by new `Remote<T>` type

### LINKED ISSUE #9027: Use `Remote<T>` to indicate tolerance of far types
## What is the Problem Being Solved?

The state-of-the-art typing used in Agoric SDK to indicate potentially remote presences is:

1. `ERef<T>` only indicates that `T` may or may not be a promise.  The downside is after `await T`, is performed, there is no propagation of the remote nature of T.
2. `FarRef<T>` is a more complex type that attempts to express promise-or-not, awaitable local properties, and a remote brand check, but without propagation.

It would be useful to have a type designation (`Remote<T>`) for an object created by `Far` or an exo, which says that a function producing or consuming such a type will (by default) treat it as potentially remote.

## Description of the Design

Now that `RemotableBrand` has been supported for a long time in `@endo/eventual-send`, and has been added to `Far`, `Remotable`, and `zone.exo*`, we can use it.  This will build more sophisticated remote typing than `ERef<T>` and clearer than `FarRef<T>`.

With care, `Remote<T>` can be used in any situation, conserving information as a function argument or return value, as well as a variable or in a data structure.

## Security Considerations

Improves correctness when handling remote objects.

## Scaling Considerations

Encourages synchronous calls to be converted to async, which introduces more turns.  For performance-sensitive code running in the same vat as a remotable such as `Remote<{ getNum(): number }>`, if a locality assertion like `'getNum' in remotable` is true, `remotable` will be narrowed to a synchronous callable (such that `remotable.getNum()` is possible).

## Test Plan

Use `Remote<T>` in existing code and verify that Typescript only points out problems to solve that are understandable consequences of the new typing.

## Upgrade Considerations

Types are erased, so there is no runtime impact on code that uses `Remote<T>`.

### PR DESCRIPTION
**Pull request by iomekam**
closes: #9027
refs: #8444

## Description

This PR creates a new `Remote<T>` type to use in code that does not assume that `T` is local, and thus must be manipulated using something like the `E` eventual send operator.  Its definition is significantly more straightforward than `FarRef`, and more effective than `ERef`.

Using the associated `Unwrap<T>` in `packages/vow`, and applying `Remote<T>` to the latest `packages/network` and `packages/vat/src/ibc.js`, I was able to surface some untypable code that indicated misuse of vows or a missing `E`.  Satisfyingly, resolving these errors led to clearer, more correct code.

### Security Considerations

Improved correctness.

### Scaling Considerations

No runtime cost.

### Documentation Considerations

Waiting until this is migrated to Endo and replaces `FarRef`, with full adoption before documenting.

### Testing Considerations

Failures are a compile-time error, but only manual testing and code linting/typechecking has been done so far.

### Upgrade Considerations

Compile-time types are erased; the definition of these types does not affect the actual code to be evaluated.

### CODE REVIEW

**Review by michaelfig — APPROVED** [review]
LGTM!

File: packages/network/src/bytes.js [hunk 1]
```diff
@@ -1,23 +1,41 @@
+// @ts-check
 /// <reference path="./types.js" />
 import { encodeBase64, decodeBase64 } from '@endo/base64';
 
+/** @typedef {Data | Buffer | Uint8Array | Iterable<number>} SourceData */
```
Comment (gibson042) [inline packages/network/src/bytes.js]: If you're going to be addressing types in this package, how about really fixing them? "Data" and "SourceData" are pretty awful names.

File: packages/network/src/bytes.js [hunk 2]
```diff
@@ -1,23 +1,41 @@
...
+  }
+  if (contents instanceof Uint8Array) {
+    // Reconstruct to ensure we have a Uint8Array and not a Buffer.
+    return new Uint8Array(
+      contents.buffer,
+      contents.byteOffset,
+      contents.byteLength,
+    );
+  }
+  return new Uint8Array(contents);
```
Comment (gibson042) [inline packages/network/src/bytes.js]: Things like `str.split("").map()` are pretty bad in current XS, cf. https://github.com/endojs/endo/issues/1984
```suggestion
  if (typeof contents === 'string') {
    return Uint8Array.from(contents, c => {
      const b = c.charCodeAt(0);
      b <= 0xFF || Fail`character cannot be coerced to an octet: ${c}`;
      return b;
    });
  } else if (contents instanceof Uint8Array) {
    // Reconstruct to ensure we have a Uint8Array and not a Buffer.
    return new Uint8Array(
      contents.buffer,
      contents.byteOffset,
      contents.byteLength,
    );
  }
  return new Uint8Array(contents);
```

File: packages/network/src/bytes.js [hunk 3]
```diff
@@ -1,23 +1,41 @@
...
+};
+
 /**
- * Convert some data to bytes.
+ * Convert some data to bytes.  We can use a local Iterable as a source for data
+ * bytes, but we need to call `toBytes(it)` to obtain a passable value.
+ *
+ * NOTE: Passing a Uint8Array without calling this function is not yet supported
+ * by `@endo/marshal`.
```
Comment (gibson042) [inline packages/network/src/bytes.js]:
```suggestion
 * Convert a Uint8Array or other sequence of octets to a string representation
 * that `@endo/marshal` accepts as Passable.
```

File: packages/network/src/bytes.js [hunk 4]
```diff
@@ -1,23 +1,41 @@
...
-  // TODO: We really need marshallable TypedArrays.
-  if (typeof bytes === 'string') {
-    bytes = bytes.split('').map(c => c.charCodeAt(0));
-  }
-
   // We return the raw characters in the lower half of
   // the String's representation.
-  const buf = new Uint8Array(bytes);
+  const buf = coerceToByteArray(data);
   return String.fromCharCode.apply(null, buf);
```
Comment (gibson042) [inline packages/network/src/bytes.js]:
```suggestion
  return String.fromCharCode(...buf);
```

**Review by gibson042 — COMMENTED** [review]
Comments from a started-but-not-finished review of #9025.

File: packages/network/src/bytes.js (hunk 1 above)
Comment (iomekam) [reply packages/network/src/bytes.js]: Any suggestions on good names?

Comment (michaelfig) [reply packages/network/src/bytes.js]: "SourceData" could become "InputBytes", and "Data" could just be replaced by the existing "Bytes".  Correctly renaming all the instances of `Data` will be annoying, but worth it.

Comment (michaelfig) [reply packages/network/src/bytes.js]: Or `SourceData` could be `ByteSource`.

**Pull request merged**